linetools/analysis/continuumfnd.py

```diff
@@ -0,0 +1,206 @@
+      green - pixels used to define continuum knots
+      black - original spectrum , gray - error in flux
+      blue line - initially defined continuum (not used),
+                  and initially defined continuum shifted by the error in flux in the smoothed spectrum (also not used)
+      red line - smoothed spectrum
+    lam0 : float
+      lowest wavelength to plot
+    dlam : float
+      wavelenght interval width to plot; highest wavelength shown will be lam0+dlam
+    yfmax : float
+      maximum ylim is yfmax * np.median(flux)
+    ymin : float
+      minimum ylim is yfmin * np.median(flux)
+    outf : str
+      output file with knots (.jsn or .json type)
+    overwrite : bool
+      overwrite?
+
+    Returns:
+    --------
+    knots : list of 2-element lists
+      List of knots (wavelength, flux) that define continuum.
+    knotpixs : list of int
+      Indices of pixels that define continuum
+
+    """
+
+    # input spectrum
+    if type(spec) is xspec.XSpectrum1D:
+        sp = spec
+    else:
+        try:
+            sp = readspec(spec)
+        except OSError:
+            print('Not valid input. Please provide XSpectrum1D object, or a file that contains XSpectrum1D object')
+            return
+
+    wav = sp.wavelength
+    flx = sp.flux
+    sigf = sp.sig
+
+    # smoothed input spectrum by smsm pixels
+    sp3 = sp.box_smooth(sm)
+    wav3 = sp3.wavelength
+    flx3 = sp3.flux
+    sigf3 = sp3.sig
+
+
+    # 1) Define pixels that are on the continuum.
+    #    Smooth the spectrum, find pixels around maximums in the flux in the smoothed spectrum,
+    #    assume that the continuum is well defined by these pixels only.
+    #    This flux will be too high, and is used only in the step (1)
+    ipixs = []  # continuum pixels
+    for i in range(len(flx3)):
+        i1 = np.max([0, i - npix])
+        i2 = np.min([i + npix, (len(flx3)) - 1])
+        ii = np.linspace(i1, i2, i2 - i1 + 1)
+        isum = 0
+        for j in ii:
+            if (flx3[int(j)] <= flx3[i]):
+                isum = isum + 1
+            if (flx3[int(j)] >= flx3[i]):
+                isum = isum - 1
+        if (isum) == len(ii) - 1:
+            ipixs.append(i)
+
+    # interpolate continuum to all pixels
+    cont0ipixs = flx3[ipixs]
+    cont0 = np.interp(wav, wav3[ipixs], cont0ipixs)
+
+
+    # 2) Identify pixels that correspond to absorption lines. Identify all other pixels.
+    #    Define continuum as in (1),
+    #    define absorption line as a set of adjacent pixels that are below the continuum level; flux of the boundary
+    #      pixels exceeds continuum level
+    #    Define EW = sum of (1 - flux/continuum) of these adjacent pixels
+    #    and a condition that EW > ewsnlim * sigma(EW) ; default ewsnlim = 2
+
+    # find pixels that correspond to absorption lines
+    ind0s = [] # pixels corresponding to absorption lines
+    imin = -1
+    iew = 0
+    isigew = 0
+
+    for i in range(len(flx)):
```

Comment:
Gack, another `for` loop. 

Surely you can use logic to do this without a `for` loop.

Reply:
Now not using for loops. Instead, using np.cumsum, np.diff, and np.where. The code is also now faster.